torchacc/ops/context_parallel/utils.py

```diff
@@ -155,28 +155,54 @@ def flash_attention(
                             cu_seqlens_q).clone()
 
 
-def slice_forward(tensor: torch.Tensor,
-                  seq_dim: int,
-                  process_group: Optional[dist.ProcessGroup] = None):
-    """ Slice the input tensor along sequence dimension into multiple chunks,
-        which are parallelized across GPUs in a context parallel group.
+def all_gather(tensor: torch.Tensor,
+               process_group: Optional[dist.ProcessGroup] = None):
+    cp_size = dist.get_world_size(process_group)
+    if cp_size == 1:
+        return [tensor]
+    tensor_list = [torch.empty_like(tensor) for _ in range(cp_size)]
+    torch.distributed.all_gather(tensor_list, tensor, group=process_group)
+    return tensor_list
 
-    Args:
-        tensor (torch.Tensor): [batch_size, seqlen, nheads, headdim].
-        seq_dim (int): The dimension to be split.
-        process_group (torch.distributed.ProcessGroup, optional): The context parallel group.
 
-    Returns:
-        tensor (torch.Tensor): [batch_size, seqlen // cp_size, nheads, headdim].
+def _split(input, dim, process_group):
+    # skip if world_size == 1
+    rank = dist.get_rank(process_group)
+    world_size = dist.get_world_size(process_group)
+    if world_size == 1:
+        return input
+
+    # split sequence
+    if input.size(dim) % world_size != 0:
+        raise ValueError(f"The seqlen {input.size(dim)} needs to" \
+                          f" be divisible by the size of process group {world_size}.")
+    return input.chunk(world_size, dim=dim)[rank].contiguous()
+
+
+def _gather(input, dim, process_group):
+    # skip if world_size == 1
+    world_size = dist.get_world_size(process_group)
+    if world_size == 1:
+        return input
+
+    # gather sequence
+    output = all_gather(input, process_group)
+    return torch.cat(output, dim=dim).contiguous()
+
+
+class SplitForwardGatherBackward(torch.autograd.Function):
+    """Split the input and scatter to context parallel region.
     """
-    cp_size = dist.get_world_size(process_group)
-    cp_rank = dist.get_rank(process_group)
-    if cp_size > 1:
-        if tensor.shape[seq_dim] % cp_size != 0:
-            raise ValueError(f"The seqlen {tensor.shape[seq_dim]} needs to" \
-                             f" be divisible by the size of process group {cp_size}.")
-        tensor = tensor.chunk(cp_size, dim=seq_dim)[cp_rank].contiguous()
-    return tensor
+
+    @staticmethod
+    def forward(ctx, tensor, seq_dim, process_group):
+        ctx.process_group = process_group
+        ctx.seq_dim = seq_dim
+        return _split(tensor, seq_dim, process_group)
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        return _gather(grad_output, ctx.seq_dim, ctx.process_group), None, None
```

Comment:
Do we need to divide by world_size here?

Reply:
add the argument 'grad_scale' to enable scale by world_size